scripts/Muon/GUI/Common/contexts/fitting_context.py

```diff
@@ -6,36 +6,153 @@
 # SPDX - License - Identifier: GPL - 3.0 +
 from __future__ import (absolute_import, division)
 
+from collections import OrderedDict
+
+from mantid.py3compat import iterkeys
+
 from Muon.GUI.Common.observer_pattern import Observable
 
+# Magic values for names of columns in the fit parameter table
+NAME_COL = 'Name'
+VALUE_COL = 'Value'
+ERRORS_COL = 'Error'
+
+
+def is_same_parameter(prefixed_name, unprefixed_name):
+    """
+    Check if two parameters are actually referring to the same parameter name.
+    :param prefixed_name: A name such as f0.Lambda from the result of a fit
+    :param unprefixed_name: A bare parameter name from a function, e.g. Lambda
+    :return: True if these refer to the same parameter, False otherwise
+    """
+    return prefixed_name.endswith(unprefixed_name)
+
+
+def _create_unique_param_lookup(parameter_workspace, global_parameters):
+    """
+    Create a dictionary with the parameter names as keys where each
+    value is a 2-tuple of (value, error). Only unique parameters are
+    included, i.e if parameters are marked global then only 1 parameter
+    is kept.
+
+    :param parameter_workspace: A raw parameter table from Fit
+    :param global_parameters: A list of global parameters(if any)
+    """
+
+    def is_in(unique_params, param_name):
+        if not global_parameters:
+            return False
+
+        # Find the global parameter than matches this one
+        global_name = None
+        for name in global_parameters:
+            if is_same_parameter(param_name, name):
+                global_name = name
+                break
+        if global_name is None:
+            # param_name is not in globals
+            return False
+
+        # Do we have this parameter already?
+        for unique_name in iterkeys(unique_params):
+            if is_same_parameter(unique_name, name):
```

Comment:
won't name be out of scope here as its only defined in the above for loop?

Reply:
Good point. This will work in Python 2 as its scoping rules are relaxed but not in 3 :+1: